blockchain/stakeversion_test.go

```diff
@@ -0,0 +1,805 @@
+		},
+		{
+			name:       "27 33 10000",
+			skip:       27,
+			interval:   33,
+			multiplier: 10000,
+		},
+		{
+			name:       "mainnet params",
+			skip:       chaincfg.MainNetParams.StakeValidationHeight,
+			interval:   chaincfg.MainNetParams.StakeVersionInterval,
+			multiplier: 5000,
+		},
+		{
+			name:       "testnet params",
+			skip:       chaincfg.TestNetParams.StakeValidationHeight,
+			interval:   chaincfg.TestNetParams.StakeVersionInterval,
+			multiplier: 1000,
+		},
+		{
+			name:       "simnet params",
+			skip:       chaincfg.SimNetParams.StakeValidationHeight,
+			interval:   chaincfg.SimNetParams.StakeVersionInterval,
+			multiplier: 10000,
+		},
+		{
+			name:       "negative mainnet params",
+			skip:       chaincfg.MainNetParams.StakeValidationHeight,
+			interval:   chaincfg.MainNetParams.StakeVersionInterval,
+			multiplier: 1000,
+			negative:   1,
+		},
+	}
+
+	for _, test := range tests {
+		t.Logf("running: %v skip: %v interval: %v",
+			test.name, test.skip, test.interval)
+
+		start := int64(test.skip + test.interval*2)
+		expectedHeight := start - 1 // zero based
+		x := int64(0) + test.negative
+		for i := start; i < test.multiplier*test.interval; i++ {
+			if x%test.interval == 0 && i != start {
+				expectedHeight += test.interval
+			}
+			wantHeight := calcWantHeight(test.skip, test.interval, i)
+
+			if wantHeight != expectedHeight {
+				if test.negative == 0 {
+					t.Fatalf("%v: i %v x %v -> wantHeight %v expectedHeight %v\n",
+						test.name, i, x, wantHeight, expectedHeight)
+				}
+			}
+
+			x++
+		}
+	}
+}
+
+func newNode(blockVersion int32, height int64, currentNode *blockNode) *blockNode {
+	// Make up a header.
+	header := &wire.BlockHeader{
+		Version: blockVersion,
+		Height:  uint32(height),
+		Nonce:   0,
+	}
+	node := newBlockNode(header, &chainhash.Hash{}, 0,
+		[]chainhash.Hash{}, []chainhash.Hash{},
+		[]uint32{})
+	node.height = height
+	node.parent = currentNode
+
+	return node
+}
+
+func TestCalcStakeVersionCorners(t *testing.T) {
+	params := &chaincfg.SimNetParams
+	genesisNode := genesisBlockNode(params)
+
+	bc := &BlockChain{
+		chainParams: params,
+	}
+
+	svh := params.StakeValidationHeight
+	interval := params.StakeVersionInterval
+
+	var currentNode *blockNode
+	currentNode = genesisNode
+	height := int64(0)
+	for i := int64(1); i <= svh; i++ {
+		node := newNode(0, i, currentNode)
+
+		// Don't set stake versions.
+
+		currentNode = node
+		bc.bestNode = currentNode
+		height = i
+	}
+	if height != svh {
+		t.Fatalf("invalid height got %v expected %v", height,
+			params.StakeValidationHeight)
+	}
+
+	// Generate 3 intervals with v2 votes and calculate StakeVersion.
+	runCount := interval * 3
+	for i := int64(0); i < runCount; i++ {
+		node := newNode(3, height+i, currentNode)
+
+		// Set stake versions.
+		for x := uint16(0); x < params.TicketsPerBlock; x++ {
+			node.voterVersions = append(node.voterVersions, 2)
+		}
+
+		sv, _ := bc.calcStakeVersionByNode(currentNode)
```

Comment:
Tests shouldn't be ignoring errors.  It makes things super hard to track down if something unexpected happens.

```Go
sv, err := bc.calcStakeVersionByNode(currentNode)
if err != nil {
	t.Fatalf("calcStakeVersionByNode: unexpected error: %v", err)
}
```